File: 贪吃蛇_重构/food.cpp
#include "Food.h"
#include "Console.h"
#include "stdlib.h"
#include "time.h"

Food::Food() :x(0), y(0), section(1)
{

}

Food::~Food()
{

}

void Food::display(bool isDisplay)
{
	Console::getInstance()->moveCursor(x, y);
	if (isDisplay)
	{
		printf("*");
	}
	else
	{
		printf(" ");
	}
}

void Food::updatePos(int up, int down, int left, int right)
{
	int newX = 0, newY = 0;
	//生成出现在地图边界之内的坐标
	//直到生成和之前的食物坐标不同的坐标为止
	while (true)
	{
		srand(time(0));
		newX = rand() % (right - left - 1) + left + 1;
		srand(time(0));
		newY = rand() % (down - up - 1) + up + 1;
		if (newX != x && newY != y)
		{
			break;
		}
	}
	x = newX;
	y = newY;
}

void Food::setSection(int section)
{
	this->section = section;
}

int Food::getSection()
{
	return section;
}

File: Snake_CSharp/Sprite/Food.cs
using System;

namespace Snake.Sprite
{
    /// <summary>
    /// 食物类
    /// </summary>
    class Food
    {
        //字段
        /// <summary>
        /// 当前食物的x坐标
        /// </summary>
        private int x;
        /// <summary>
        /// 当前食物的y坐标
        /// </summary>
        private int y;

        //属性
        /// <summary>
        /// 当前食物的x坐标
        /// </summary>
        public int X
        {
            get { return x; }
        }
        /// <summary>
        /// 当前食物的y坐标
        /// </summary>
        public int Y
        {
            get { return y; }
        }
        //这个属性没有定义字段，这个属性包含了get和set的两个方法，而且C#在后台为这个属性自动创建了字段，所有的方法都对这个默认字段操作
        /// <summary>
        /// 蛇吃到了这个食物之后增长的节数
        /// </summary>
        public int BodyGrowNum { get; set; }

        /// <summary>
        /// 构造函数
        /// </summary>
        /// <param name="x">初始的x坐标</param>
        /// <param name="y">初始的y坐标</param>
        /// <param name="BodyGrowNum">吃到这个食物的时候蛇增长节数</param>
        public Food(int x = 0, int y = 0, int BodyGrowNum = 1)
        {
            //初始化食物相关属性
            this.x = x;
            this.y = y;
            this.BodyGrowNum = BodyGrowNum;
        }

        //方法
        /// <summary>
        /// 是否显示食物坐标
        /// </summary>
        /// <param name="isDisplay"></param>
        public void display(bool isDisplay)
        {
            try
            {
                Console.SetCursorPosition(x, y);
            }
            catch (ArgumentOutOfRangeException) //倘若坐标超出界限就停止这一次的食物显示或者隐藏
            {
                return;
            }
            if (isDisplay) //显示/隐藏食物
            {
                Console.Write('*');
            }
            else
            {
                Console.Write(' ');
            }
        }

        /// <summary>
        /// 生成一个新的食物坐标
        /// </summary>
        public void getNewPos(int left, int right, int up, int down)
        {
            Random random = new Random(); //获取随机类
            int newX = random.Next(left + 1, right); //生成新的x坐标
            while (newX == x) //避免新的x坐标和现在的x坐标重复
            {
                newX = random.Next(left + 1, right);
            }
            int newY = random.Next(up + 1, down); //生成新的y坐标
            while (newY == y) //避免新的y坐标和现在的y坐标重复
            {
                newY = random.Next(up + 1, down);
            }
            //更新坐标
            x = newX;
            y = newY;
        }
    }
}
